src/wallet/rpcdump.cpp

```diff
@@ -805,35 +805,190 @@ UniValue dumpwallet(const JSONRPCRequest& request)
     return reply;
 }
 
+struct UsageMaps {
+    using ScriptUseMap = std::map<CScriptID, bool>;
+    using WatchKeyUseMap = std::map<CKeyID, bool>;
+    using KeyUseMap = std::map<CKeyID, bool>;
+
+    WatchKeyUseMap map_watch_key_use;
+    ScriptUseMap map_script_use;
+    KeyUseMap map_key_use;
+};
+
+// Class which remembers whether every script and public key was used
+class AllUsedKeyStore : public CBasicKeyStore
+{
+private:
+    struct UsageMaps* usage = new UsageMaps();
+public:
+    bool AllUsed() const;
+    bool GetPubKey(const CKeyID &address, CPubKey& vchPubKeyOut) const override;
+    bool AddCScript(const CScript& redeemScript) override;
+    bool GetCScript(const CScriptID &hash, CScript& redeemScriptOut) const override;
+    bool AddWatchOnly(const CScript &dest) override;
+    bool AddKeyPubKey(const CKey& key, const CPubKey &pubkey) override;
+    bool GetKey(const CKeyID &address, CKey &keyOut) const override;
+};
+
+bool AllUsedKeyStore::GetPubKey(const CKeyID &address, CPubKey &pubkey_out) const
+{
+    LOCK(cs_KeyStore);
+    if (CBasicKeyStore::GetPubKey(address, pubkey_out)) {
+        usage->map_watch_key_use[address] = true;
```

Comment:
You could invert the logic and keep track of the unused
 - when adding a key, add to the unused set
 - when reading a key, remove from the unused set

Then `AllUsed` could look like (s/usage/unused):
```cpp
    return 
        unused->watch_keys.empty() &&
        unused->scripts.empty() &&
        unused->keys.empty();
```

I could be missing some detail.